### ISSUE TITLE
Framerate < 1 fps, display of out of date frames, in example code "Using NetGear_Async with Variable Parameters"

### ISSUE THREAD

**Issue by whogben** [abhiTronix/vidgear#137]
## Description

The client for the example [Using NetGear_Async with Variable Parameters](https://abhitronix.github.io/vidgear/gears/netgear_async/usage/) does:
1. display the first frame from the server on connection rather than the most recent
2. display frames updating at a frame rate < 1 per second

### Acknowledgment

- [x] I have searched the [issues](https://github.com/abhiTronix/vidgear/issues) for my issue and found nothing related or helpful.
- [x] I have read the [Documentation](https://abhitronix.github.io/vidgear).
- [x] I have read the [Issue Guidelines](https://abhitronix.github.io/contribution/issue/).
### Environment

Server:
Python 3.8
Raspberry Pi 4 4Gb
Current Raspbian
Current test branch of vidgear installed today (however issue occurred with regular pip3 branch too)
Running on a local network, (other video-streaming approaches do work at expected frame rates, seemingly implying it's not the network)

Client:
Mac OS Catalina
Python 3.8
Current pip3 install vidgear[asyncio]

### Expected Behavior

Whether the server or the client starts first, the client will display the latest frame from the server with reasonable frame rate (eg, > 5 frames per second).

### Actual Behavior

When the server starts first, then when the client connects, it starts showing frames one by one from when the server started at an extremely low rate (frame rate < 1 per second).

When the client starts first, and then the server starts, the client does start with the servers first (current) frame, but the frame rate on the client is so slow as to quickly get behind (< 1 frame per second)

### Possible Fix

Maybe there is config required to make the client/server ignore past frames, and start from the most recent frame when they connect?

Also, this warning prints out when the server starts:
`CamGear :: WARNING :: Threaded Queue Mode is disabled for the current video source!`

If I wave my arms like a madman while starting the server, and then start the client, I can see my arms waving in super super slow motion. That indicates to me that the server is in fact taking frames at the right frame rate (or else the arm wave would be jerkier), and the issue may be more with the client side, or in the transport of frames.

### Steps to reproduce

1. On the server, make sure to get the test branch of vidgear[asyncio] as described here: https://abhitronix.github.io/vidgear/installation/source_install/#installation

2. On the server, setup the example code "Using NetGear_Async with Variable Parameters" and run it:
```
# import libraries
from vidgear.gears.asyncio import NetGear_Async
import asyncio

#initialize Server with suitable source
server=NetGear_Async(source=0, address='192.168.1.14', port='5454', protocol='tcp',  pattern=3, logging=True).launch()

if __name__ == '__main__':
    #set event loop
    asyncio.set_event_loop(server.loop)
    try:
        #run your main function task until it is complete
        server.loop.run_until_complete(server.task)
    except KeyboardInterrupt:
        #wait for keyboard interrupt
        pass
    finally:
        # finally close the server
        server.close()
```

3. On the client, setup the code from the same example and run it:

```
# import libraries
from vidgear.gears.asyncio import NetGear_Async
import cv2, asyncio

#define and launch Client with `receive_mode=True`. #change following IP address '192.168.x.xxx' with yours
client=NetGear_Async(address='192.168.1.14', port='5454', protocol='tcp',  pattern=3, receive_mode=True, logging=True).launch()


#Create a async function where you want to show/manipulate your received frames
async def main():
    # loop over Client's Asynchronous Frame Generator
    async for frame in client.recv_generator():


        # do something with received frames here
        

        # Show output window
        cv2.imshow("Output Frame", frame)
        key=cv2.waitKey(1) & 0xFF

        #await before continuing
        await asyncio.sleep(0.00001)


if __name__ == '__main__':
    #Set event loop to client's
    asyncio.set_event_loop(client.loop)
    try:
        #run your main function task until it is complete
        client.loop.run_until_complete(main())
    except KeyboardInterrupt:
        #wait for keyboard interrupt
        pass

    # close all output window
    cv2.destroyAllWindows()
    # safely close client
    client.close()
```

**Comment by whogben** [comment]
Attaching a video showing what I mean:
[trimmed.mov.zip](https://github.com/abhiTronix/vidgear/files/4775283/trimmed.mov.zip)

I started the server, and a few seconds after it ran, brought my arms up and wiggled them.

When I then started the client, I saw only myself sitting (no arms up). I closed the client, re-opened it about 10 minutes later, and at that point the client showed my arms up, and you can see the relative speed issue from the wiggling of the arms, and someone else's hand coming into frame at the end.

Based on this, I am pretty confident all the frames are taken by the server, and then getting cached somewhere, and read out very slowly from the client.

The behavior I'm trying to achieve is: when the client connects, all frames that are older than the connection would be ignored, and it would start on the most current frame. Further, it would keep displaying frames at the rate they're being sent. I can solve the out of date starting frames by starting the server only after the client is started, but the low frame rate still results in it immediately going out of date.

**Comment by abhiTronix** [comment]
> extremely low rate (frame rate < 1 per second).

@whogben Thanks for this detailed info. I have few questions like, Is something is wrong with your Network/router/wifi setting, ans are you using any proxy settings for your network? Also, are you using connected camera for any other purpose/task while running these examples?

> **Maybe there is config required to make the client/server ignore past frames, and start from the most recent frame when they connect?
> 
> Also, this warning prints out when the server starts:
> CamGear :: WARNING :: Threaded Queue Mode is disabled for the current video source!**
> 

No, this a normal warning in [Threaded Queue Mode](https://abhitronix.github.io/vidgear/bonus/TQM/) which is disabled for live source, and has nothing to do with NetGear_Async API.

## Possible Solutions

Please follow following steps:

### Step 1: Update latest ZMQ manually on both machines:

1. On macOS machine, do this:
```sh
# uninstall zmq
pip3 uninstall pyzmq -y

# brew install zmq
brew install zmq

# reinstall pyzmq
pip3 install -U pyzmq
```

2. Then, On Raspberry Pi
```sh
# uninstall zmq
pip3 uninstall pyzmq -y

# apt install zmq
sudo apt update && sudo apt install libzmq3-dev

# reinstall pyzmq
pip3 install -U pyzmq
```


### Step-2: On Raspberry Pi, Run following command
```sh
iw wlan0 set power_save off   # disables powersaver setting for wifi
``` 

### Step-3 Reinstall VidGear's `testing` branch:

Just now, I've pushed some commits to `testing` branch for increasing performance, so you can re-install it again by following [these instructions](https://abhitronix.github.io/vidgear/installation/source_install/#installation).

### Step-4 Run following Modified example:

Now run following modified example, instead of previous one:

### Client's End

Open a terminal on Client System _(where you want to display the input frames received from the Server)_ and execute the following python code: 

> "Note down the IP-address of this system(required at Server's end) by executing the command: `hostname -I` and also replace it in the following code."

> "Client will throw TimeoutError if it fails to connect to the Server in given [`timeout`](../params/#timeout) value!"

> "You can terminate client anytime by pressing **`[Ctrl+C]/[⌘+C]`** on your keyboard!"

```python
# import libraries
from vidgear.gears.asyncio import NetGear_Async
import cv2, asyncio

#define and launch Client with `receive_mode=True`. #change following IP address '192.168.x.xxx' with yours
client=NetGear_Async(address='192.168.x.xxx', port='5454', protocol='tcp',  pattern=2, receive_mode=True, logging=True).launch()


#Create a async function where you want to show/manipulate your received frames
async def main():
    # loop over Client's Asynchronous Frame Generator
    async for frame in client.recv_generator():

        
        # do something with received frames here

        
        # Show output window
        cv2.imshow("Output Frame", frame)
        key=cv2.waitKey(1) & 0xFF

        #await before continuing
        await asyncio.sleep(0.00001)


if __name__ == '__main__':
    #Set event loop to client's
    asyncio.set_event_loop(client.loop)
    try:
        #run your main function task until it is complete
        client.loop.run_until_complete(main())
    except KeyboardInterrupt:
        #wait for keyboard interrupt
        pass
        
    # close all output window
    cv2.destroyAllWindows()
    # safely close client
    client.close()
```

### Server's End

Now, Open the terminal on another Server System _(with a webcam connected to it at index `0`)_, and execute the following python code: 

> "Replace the IP address in the following code with Client's IP address you noted earlier."

> "You can terminate stream on both side anytime by pressing **`[Ctrl+C]/[⌘+C]`** on your keyboard!"

```python
# import library
from vidgear.gears.asyncio import NetGear_Async
from vidgear.gears.asyncio.helper import reducer
import cv2, asyncio

#initialize Server
server=NetGear_Async(source=None, address='192.168.x.xxx', port='5454', protocol='tcp',  pattern=2, logging=True)

#Create a async frame generator as custom source
async def my_frame_generator():

        #Open any video stream such as live webcam video stream on first index(i.e. 0) device
        stream=cv2.VideoCapture(0) 

        # loop over stream until its terminated
        while True:

            # read frames
            (grabbed, frame)=stream.read()

            # check if frame empty
            if not grabbed:
                #if True break the infinite loop
                break

            # reducer frames size if you want more performance, otherwise comment this line
            frame = reducer(frame, percentage = 30) #reduce frame by 30%

            # do something with the frame to be sent here

            # yield frame
            yield frame
            # sleep for sometime
            await asyncio.sleep(0.00001)


if __name__ == '__main__':
    #set event loop
    asyncio.set_event_loop(server.loop)
    #Add your custom source generator to Server configuration
    server.config["generator"]=my_frame_generator() 
    #Launch the Server 
    server.launch()
    try:
        #run your main function task until it is complete
        server.loop.run_until_complete(server.task)
    except KeyboardInterrupt:
        #wait for keyboard interrupt
        pass
    finally:
        # finally close the server
        server.close()
```

and see if this boost performance and then report your results here. Goodluck!

**Comment by whogben** [comment]
Wow thanks for all that!

**Question Responses**
- I don't control the network, but I don't think it's running any proxies. It is a verizon fios modem assigning IPs, connected via ethernet to a netgear router that the machines are actually connected to. The RPi is connected to the 2.4ghz version while the Mac is on the 5ghz version. It's possible traffic is routed out to the Verizon modem and then back into the netgear before reaching me - but other video solutions I've been testing have not experienced as much trouble, and VNC streams the screen well.
- I don't believe any other process is using the camera.

### Suggested Tests Results:

1) Rerunning original setup:
- when server starts first, same behavior as original issue
- when client starts first:, the client gives the following output:
```
NetGear_Async :: DEBUG :: Launching NetGear asynchronous generator!
NetGear_Async :: DEBUG :: Successfully Binded to address: tcp://192.168.1.14:5454 with pattern: 3.
NetGear_Async :: DEBUG :: Receive Mode is activated successfully!
Traceback (most recent call last):
  File "vidgeartest.py", line 31, in <module>
    client.loop.run_until_complete(main())
  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
  File "vidgeartest.py", line 12, in main
    async for frame in client.recv_generator():
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vidgear/gears/asyncio/netgear_async.py", line 363, in recv_generator
    recvd_msg = await asyncio.wait_for(
  File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
    raise exceptions.TimeoutError()
asyncio.exceptions.TimeoutError
```

2) After reinstalling ZMQ on Mac and Pi (old version on both was 19.0.1,  new version installed was also 19.01)
- when server starts first:  same behavior as original issue
- when client starts first:  same error as test 1

***unexpected behavior:***  I stopped the server on the raspberry pi, and ran the client on the mac again (in preparation to start the server again), even though the server was closed, out of date frames were displayed.  I then started the server again, and after a short time, a new up to date frame was displayed - though the speed stayed <1fps and got out of date again.  It's as though the frames were buffered somewhere, but restarting the server cut through that and sent a new one over the top.

3) reinstalling vidgear testing branch (on Mac and RPi):
- running server first: same original delay behavior
- running client first: same error as test 1 & 2

4) Running modified example:
- Running the server code gives this output:
```
NetGear_Async :: WARNING :: Given source is of NoneType!
NetGear_Async :: DEBUG :: Creating NetGear asynchronous server handler!
NetGear_Async :: DEBUG :: Successfully connected to address: tcp://192.168.1.14:5454 with pattern: 2.
NetGear_Async :: DEBUG :: Send Mode is successfully activated and ready to send data!
NetGear_Async :: DEBUG :: Terminating various Send Mode Processes.
Traceback (most recent call last):
  File "/home/pi/Desktop/netgear_example.py", line 46, in <module>
    server.loop.run_until_complete(server.task)
  File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
  File "/home/pi/.local/lib/python3.7/site-packages/vidgear/gears/asyncio/netgear_async.py", line 299, in __server_handler
    if not (frame.flags["C_CONTIGUOUS"]):
AttributeError: 'coroutine' object has no attribute 'flags'
```

### Additional Tests:

1) Try running both client and server on the pi together, using original example code, hopefully to rule out the network / router setup as an issue:
- when client starts first:  same timeout error as previous tests
- when server starts first: the behavior is slightly less slow, but it still took 18 seconds to see my hand appear after I moved it in front of the camera

2) Try running both modified examples on the pi together:
- client first, timeout error
- server first, same 'coroutine' object has no attribute 'flags' error

3) Try running both client and server on the mac together, using original example code:
- client first, timeout error
- server first, took 24 seconds to see my hand in the video after I put it in front of the camera

### TLDR:
- probably ruled out the network being the issue: 18 and 24 second delays when running locally to 127.0.0.1, 
- discovered I'm using python 3.7 on the pi not 3.8 as I thought, not sure if that's considered to be a possible problem.
- General question: Is it working / streaming smoothly for you when you run the examples? If so, what computer/config?

Anyway, thank you so much for all the great help with this - I am really really excited to use vidgear to get low latency streaming from the pi - this is an amazing framework and your response has been amazing too!

**Comment by abhiTronix** [comment]
> Rerunning original setup:
> when server starts first, same behavior as original issue
> when client starts first:, the client gives the following output:
> NetGear_Async :: DEBUG :: Launching NetGear asynchronous generator!
> NetGear_Async :: DEBUG :: Successfully Binded to address: tcp://192.168.1.14:5454 with pattern: 3.
> NetGear_Async :: DEBUG :: Receive Mode is activated successfully!
> Traceback (most recent call last):
>   File "vidgeartest.py", line 31, in <module>
>     client.loop.run_until_complete(main())
>   File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
>   File "vidgeartest.py", line 12, in main
>     async for frame in client.recv_generator():
>   File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/site-packages/vidgear/gears/asyncio/netgear_async.py", line 363, in recv_generator
>     recvd_msg = await asyncio.wait_for(
>   File "/Library/Frameworks/Python.framework/Versions/3.8/lib/python3.8/asyncio/tasks.py", line 490, in wait_for
>     raise exceptions.TimeoutError()
> asyncio.exceptions.TimeoutError

@whogben this TimeoutError can be avoided  by setting a high value  on [`timeout`](https://abhitronix.github.io/vidgear/gears/netgear_async/params/#timeout) parameter at receiver's end. But problem is somewhere else, maybe in your network, since I'm not able to reproduce this much latency <1fps on my Raspberry pi 3B and Windows computer connected over 2.4ghz Wi-fi. **Are you using USB Camera instead of [Pi Camera Module](https://www.raspberrypi.org/products/camera-module-v2/)?** 

> After reinstalling ZMQ on Mac and Pi (old version on both was 19.0.1, new version installed was also 19.01)
> when server starts first: same behavior as original issue
> when client starts first: same error as test 1

Re-installing ZMQ binaries from apt or brew does helps, according my tests. But again it's something else in your case that is throttling the performance.


> unexpected behavior: I stopped the server on the raspberry pi, and ran the client on the mac again (in preparation to start the server again), even though the server was closed, out of date frames were displayed. I then started the server again, and after a short time, a new up to date frame was displayed - though the speed stayed <1fps and got out of date again. It's as though the frames were buffered somewhere, but restarting the server cut through that and sent a new one over the top.

**Reason:** You can see if it will block or drop by looking through the documentation for the socket type for [ZMQ::HWM option action](http://zeromq.github.io/rbzmq/classes/ZMQ/Socket.src/M000006.html) which will either be Block or Drop. The default ZMQ::HWM value of zero means “no limit”, meaning that it will not enter an exceptional state until the system runs out of memory, which is happening your case.

**Solution:** I've pushed the update to set ZMQ::HWM value automatically in `testing` branch, that;s why I asked you to reinstall it from scratch.

---

> Running modified example:
> Running the server code gives this output:
> NetGear_Async :: WARNING :: Given source is of NoneType!
> NetGear_Async :: DEBUG :: Creating NetGear asynchronous server handler!
> NetGear_Async :: DEBUG :: Successfully connected to address: tcp://192.168.1.14:5454 with pattern: 2.
> NetGear_Async :: DEBUG :: Send Mode is successfully activated and ready to send data!
> NetGear_Async :: DEBUG :: Terminating various Send Mode Processes.
> Traceback (most recent call last):
>   File "/home/pi/Desktop/netgear_example.py", line 46, in <module>
>     server.loop.run_until_complete(server.task)
>   File "uvloop/loop.pyx", line 1456, in uvloop.loop.Loop.run_until_complete
>   File "/home/pi/.local/lib/python3.7/site-packages/vidgear/gears/asyncio/netgear_async.py", line 299, in __server_handler
>     if not (frame.flags["C_CONTIGUOUS"]):
> AttributeError: 'coroutine' object has no attribute 'flags'

Apologies, I've missed a `await` in above example. **Anyways, download following file:**

[netgear_async test.zip](https://github.com/abhiTronix/vidgear/files/4778586/netgear_async.test.zip)

It contains proper working code for both Server and Client end. And, **You need to do following task with the files contained in it:**

0. ⚠️ Important : Install `testing` branch again.
1. The files in `local` folder must be run in different terminals on same machine _(i.e server on terminal-1 and client on terminal-2)_, make sure you have your camera connected to your this machine _(at index `0`)_.
2. The files in `multiple` folder must be run on different machine(Pi and MacOS), like above. Please make sure you change the `address` parameter value to yours.
3. Post result here.

----

> Try running both client and server on the pi together, using original example code, hopefully to rule out the network / router setup as an issue:
> when client starts first: same timeout error as previous tests

Meaning, your are letting your Client alone for long time without Server, thus, triggering this Timeout error. 

> when server starts first: the behavior is slightly less slow, but it still took 18 seconds to see my hand appear after I moved it in front of the camera

This is unbelievable, there is something wrong with your machine. Are you using [this bare-minimum code](https://abhitronix.github.io/vidgear/gears/netgear_async/usage/#bare-minimum-usage) to test the performance on same machine?

See my results on much less powerful Raspberry Pi 3B than your Raspberry Pi 4:

[2020-06-15-12-48-03.zip](https://github.com/abhiTronix/vidgear/files/4778613/2020-06-15-12-48-03.zip)

and I'm recording VNC output which is usually slower than in real.

> Try running both client and server on the mac together, using original example code:
> client first, timeout error
> server first, took 24 seconds to see my hand in the video after I put it in front of the camera

That's weird because You can see exact real-time output on my Windows 10 Machine:

[2020-06-15-12-34-05.zip](https://github.com/abhiTronix/vidgear/files/4778522/2020-06-15-12-34-05.zip)

> TLDR:
> probably ruled out the network being the issue: 18 and 24 second delays when running locally to 127.0.0.1,
> discovered I'm using python 3.7 on the pi not 3.8 as I thought, not sure if that's considered to be a possible problem.

Nope. It maybe these problems. **Can you test [NetGear API](https://abhitronix.github.io/vidgear/gears/netgear/usage/#using-netgear-with-variable-parameters) too, to cross-check everything?**

 >General question: Is it working / streaming smoothly for you when you run the examples? If so, what computer/config?

I'm working with AMD laptop and Raspberry Pi 3B with 2.4ghz standard Wi-Fi as local network.

**Comment by abhiTronix** [comment]
Closed and marked invalid for now, as it can't be reproduced. Kindly share your results when you're ready. Good luck!

**Comment by whogben** [comment]
Under testing branch, using the provided files (thank you!):
- local test works great on mac and raspberry pi
- multiple test has the same issue

I will be by a different wifi network and able to try something new on Monday, will report back when I know more.

**Comment by abhiTronix** [comment]
>Under testing branch, using the provided files (thank you!):
  >  local test works great on mac and raspberry pi

@whogben Thanks for confirming.

> multiple test has the same issue
> I will be by a different wifi network and able to try something new on Monday, will report back when I know more.

Good approach, please do notify the results here, as It will be helpful for the community too.

**@whogben Also, try [NetGear API examples](https://abhitronix.github.io/vidgear/gears/netgear/usage) too and see if it is any faster on your machine than NetGear_Async API.**

**Comment by whogben** [comment]
Testing non async approach seems to be much much faster, 1.5 - 2 seconds latency, 2-3 frames per second, and it doesn't get further behind over time - the latency stays about the same from connecting to several minutes later. Not sure if it's relevant, but the server constantly prints out:
`NetGear :: DEBUG :: b'Data received on device: 24404 !'`

When I added a printout on the server side to the loop, showing when it captures a frame, the frame rate it's sending is about 1FPS:

```
# import required libraries
from vidgear.gears import VideoGear
from vidgear.gears import NetGear

# define various tweak flags
options = {'flag' : 0, 'copy' : False, 'track' : False}

# Open live video stream on webcam at first index(i.e. 0) device
stream = VideoGear(source=0).start()

# Define Netgear server at given IP address and define parameters (!!! change following IP address '192.168.x.xxx' with client's IP address !!!)
server = NetGear(address = '192.168.1.14', port = '5454', protocol = 'tcp',  pattern = 1, logging = True, **options)

# loop over until KeyBoard Interrupted
while True:

  try: 

    # read frames from stream
    frame = stream.read()

    # check for frame if Nonetype
    if frame is None:
        break

    # {do something with the frame here}
    print('frame')

    # send frame to server
    server.send(frame)

  except KeyboardInterrupt:
    break

# safely close video stream
stream.stop()

# safely close server
server.close()
```

My theory of the issue is now that it's the network. The server should be able to take frames much faster than 1 fps - it must be getting stuck waiting during the send() for acknowledgement from the client. As compared to asyncio version, where it likely queued up loads of frames (server showed sending rapidly) and then the network somehow buffered / delayed them down to the low rate observed. I will post again when I have the opportunity to test on new network hardware.

**Comment by abhiTronix** [comment]
> Not sure if it's relevant, but the server constantly prints out:
> NetGear :: DEBUG :: b'Data received on device: 24404 !'

@whogben just turn the logging off ([`logging = False`](https://abhitronix.github.io/vidgear/gears/netgear/params/#logging)).

> My theory of the issue is now that it's the network. The server should be able to take frames much faster than 1 fps - it must be getting stuck waiting during the send() for acknowledgement from the client. As compared to asyncio version, where it likely queued up loads of frames (server showed sending rapidly) and then the network somehow buffered / delayed them down to the low rate observed. I will post again when I have the opportunity to test on new network hardware.

Yes, indeed its network, otherwise  others would have reported this behavior too. According to my tests, NetGear_Async is sometimes faster, and other times its slower than NetGear. I don't actually know what causes this but its certain that NetGear_Async is far memory efficient than  NetGear due to its asyncio structure. I'll advise you to stick with NetGear as it worked much better for you, but do try different [`pattern`](https://abhitronix.github.io/vidgear/gears/netgear/params/#pattern) values, as each of them suits different settings.


**On top of this see [FAQ question](https://abhitronix.github.io/vidgear/help/netgear_faqs/#why-netgear-is-slow).**

**Comment by whogben** [comment]
I can confirm that changing to a new router resolved the issue, thank you so much for the help in troubleshooting this and I'm sorry it turned out to be something so simple!

For others who might be having a similar problem in the future and reading this issue - previously, the WiFi was acting in bridge mode to a second router was doing the DHCP work. My guess is something about this network topology or the second router (a verizon modem) was causing variable latency spikes, and would get backed up causing the delays described. Moving to another router setup, where a different router both provides DHCP and the WiFi network resolved the issue for me. Thanks again abhiTronix!

**Comment by abhiTronix** [comment]
@whogben **Kudos 🍻**
